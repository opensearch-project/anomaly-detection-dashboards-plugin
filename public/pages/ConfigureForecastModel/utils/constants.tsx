/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import {
  ModelConfigurationFormikValues,
} from '../../ConfigureForecastModel/models/interfaces';
import { DEFAULT_SHINGLE_SIZE } from '../../../utils/constants';

// FIXME: We intentionally leave some parameters undefined to encourage customers to click 
// the "Suggest parameters" button rather than providing default values like "10 minutes" for interval
export const INITIAL_MODEL_CONFIGURATION_VALUES: ModelConfigurationFormikValues =
  {
    shingleSize: DEFAULT_SHINGLE_SIZE,
    imputationOption: undefined,
    interval: undefined,
    windowDelay: undefined,
    suggestedSeasonality: undefined,
    recencyEmphasis: undefined,
    resultIndex: undefined,
    resultIndexMinAge: 7,
    resultIndexMinSize: 51200,
    resultIndexTtl: 60,
    flattenCustomResultIndex: false,
  };

// an enum for the sparse data handling options
export enum SparseDataOptionValue {
  IGNORE = 'ignore',
  PREVIOUS_VALUE = 'previous_value',
  SET_TO_ZERO = 'set_to_zero',
  CUSTOM_VALUE = 'custom_value',
}
