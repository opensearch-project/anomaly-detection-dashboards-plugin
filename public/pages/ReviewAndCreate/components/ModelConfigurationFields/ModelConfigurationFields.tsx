/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React, { useState } from 'react';
import {
  EuiBasicTable,
  EuiLink,
  EuiSmallButton,
  EuiSpacer,
  EuiCallOut,
  EuiLoadingSpinner,
  EuiFlexGroup,
  EuiText,
} from '@elastic/eui';
import {
  Detector,
  FEATURE_TYPE,
  FeatureAttributes,
  ValidationModelResponse,
} from '../../../../models/interfaces';
import { get, sortBy, isEqual } from 'lodash';
import ContentPanel from '../../../../components/ContentPanel/ContentPanel';
import { CodeModal } from '../../../../components/CodeModal/CodeModal';
import { AdditionalSettings } from '../AdditionalSettings/AdditionalSettings';
import { getTitleWithCount } from '../../../../utils/utils';
import {
  getShingleSizeFromObject,
  imputationMethodToFormik,
  getCustomValueStrArray,
 } from '../../../ConfigureModel/utils/helpers';
import { SORT_DIRECTION } from '../../../../../server/utils/constants';

interface ModelConfigurationFieldsProps {
  detector: Detector;
  onEditModelConfiguration(): void;
  validationFeatureResponse: ValidationModelResponse;
  validModel: boolean;
  validationError: boolean;
  isLoading: boolean;
  isCreatingDetector: boolean;
}

interface ModelConfigurationFieldsState {
  showCodeModel: boolean[];
  sortField: string;
  sortDirection: SORT_DIRECTION;
}

export const ModelConfigurationFields = (
  props: ModelConfigurationFieldsProps
) => {
  const [featuresState, setFeaturesState] =
    useState<ModelConfigurationFieldsState>({
      showCodeModel: get(props.detector, 'featureAttributes', []).map(
        () => false
      ),
      sortField: 'name',
      sortDirection: SORT_DIRECTION.ASC,
    });

  const closeModal = (index: number) => {
    const cloneShowCodeModal = [...featuresState.showCodeModel];
    cloneShowCodeModal[index] = false;
    setFeaturesState({
      ...featuresState,
      showCodeModel: cloneShowCodeModal,
    });
  };

  const showModal = (index: number) => {
    const cloneShowCodeModal = [...featuresState.showCodeModel];
    cloneShowCodeModal[index] = true;
    setFeaturesState({
      ...featuresState,
      showCodeModel: cloneShowCodeModal,
    });
  };

  const getModalVisibilityChange = (index: number) => {
    return featuresState.showCodeModel[index];
  };

  const handleTableChange = (props: any) => {
    setFeaturesState({
      ...featuresState,
      sortField: props.sort.field,
      sortDirection: props.sort.direction,
    });
  };

  const getSortedItems = (items: Array<any>) => {
    let sorted = sortBy(items, featuresState.sortField);
    if (featuresState.sortDirection == SORT_DIRECTION.DESC) {
      sorted = sorted.reverse();
    }
    return sorted;
  };
  const featureAttributes = get(props.detector, 'featureAttributes', []);
  const shingleSize = getShingleSizeFromObject(props.detector);
  const imputationMethodStr = imputationMethodToFormik(props.detector);

  const sorting = {
    sort: {
      field: featuresState.sortField,
      direction: featuresState.sortDirection,
    },
  };

  const items = featureAttributes.map(
    (feature: FeatureAttributes, index: number) => ({
      name: feature.featureName,
      definition: index,
      state: feature.featureEnabled ? 'Enabled' : 'Disabled',
    })
  );

  const sortedItems = getSortedItems(items);

  const columns = [
    {
      field: 'name',
      name: 'Feature name',
      sortable: true,
    },
    {
      field: 'definition',
      name: 'Feature definition',
      render: (featureIndex: number) => {
        const feature = featureAttributes[featureIndex];

        const metaData = get(
          props.detector,
          `uiMetadata.features.${feature.featureName}`,
          {}
        );

        if (
          Object.keys(metaData).length === 0 ||
          metaData.featureType == FEATURE_TYPE.CUSTOM
        ) {
          return (
            <div>
              <p>
                Custom expression:{' '}
                <EuiLink
                  data-test-subj={`viewFeature-${featureIndex}`}
                  onClick={() => showModal(featureIndex)}
                >
                  View code
                </EuiLink>
              </p>

              {!getModalVisibilityChange(featureIndex) ? null : (
                <CodeModal
                  code={JSON.stringify(feature.aggregationQuery, null, 4)}
                  title={feature.featureName}
                  subtitle="Custom expression"
                  closeModal={() => closeModal(featureIndex)}
                  getModalVisibilityChange={() =>
                    getModalVisibilityChange(featureIndex)
                  }
                />
              )}
            </div>
          );
        } else {
          return (
            <div>
              <p>Field: {metaData.aggregationOf || ''}</p>
              <p>Aggregation method: {metaData.aggregationBy || ''}</p>
            </div>
          );
        }
      },
    },
    {
      field: 'state',
      name: 'Feature state',
    },
  ];

  const getCellProps = () => {
    return {
      textOnly: true,
    };
  };

  const getFeatureValidationCallout = (
    issueResponse: ValidationModelResponse
  ) => {
    if (issueResponse != undefined) {
      if (issueResponse.sub_issues != undefined) {
        const renderList = Object.keys(issueResponse.sub_issues).map((key) => {
          return (
            //@ts-ignore
            <li>{'The "' + key + '" ' + issueResponse.sub_issues[key]}</li>
          );
        });
        return <ul>{renderList}</ul>;
      } else {
        return (
          <ul>
            <li>{JSON.stringify(props.validationFeatureResponse.message)}</li>
          </ul>
        );
      }
    } else {
      return null;
    }
  };
  const getValidationCallout = () => {
    //When validation response is loading then displaying loading spinner, don't display
    // after clicking on "create detector" button as isLoading will be true from that request
    if (props.isLoading && !props.isCreatingDetector) {
      return (
        <EuiCallOut
          title={
            <div>
              <EuiFlexGroup direction="row" gutterSize="xs">
                <EuiLoadingSpinner size="l" style={{ marginRight: '12px' }} />
                <EuiText>
                  <p>Validating model configurations</p>
                </EuiText>
              </EuiFlexGroup>
            </div>
          }
          style={{ marginBottom: '10px' }}
          size="s"
          color="primary"
        />
      );
    }
    if (props.validationError) {
      return null;
    }
    if (props.validModel) {
      return (
        <EuiCallOut
          title="Model configurations are validated"
          color="success"
          iconType="check"
          size="s"
          style={{ marginBottom: '10px' }}
        ></EuiCallOut>
      );
      // makes sure there is a response to display and model configs aren't valid per
      // validation API
    }
    if (
      !props.validModel &&
      props.validationFeatureResponse.hasOwnProperty('message')
    ) {
      if (
        isEqual(
          get(props, 'validationFeatureResponse.validationType', ''),
          'model'
        )
      ) {
        return (
          <EuiCallOut
            title="We identified some areas that might improve your model"
            color="warning"
            iconType="iInCircle"
            size="s"
            style={{ marginBottom: '10px' }}
          >
            {JSON.stringify(props.validationFeatureResponse.message).replace(
              /\"/g,
              ''
            )}
          </EuiCallOut>
        );
      } else {
        return (
          <EuiCallOut
            title="Issues found in the model configuration"
            color="danger"
            iconType="alert"
            size="s"
            style={{ marginBottom: '10px' }}
          >
            {/* Callout can either display feature subissue which related to a specific
            feature issue or display a feature_attribute issue that is general like
            more then x anomaly feature or dulicate feature names */}
            {props.validationFeatureResponse.hasOwnProperty('sub_issues') ? (
              getFeatureValidationCallout(props.validationFeatureResponse)
            ) : (
              <ul>
                <li>
                  {JSON.stringify(props.validationFeatureResponse.message)}
                </li>
              </ul>
            )}
          </EuiCallOut>
        );
      }
    } else {
      return null;
    }
  };
  const featureNum = Object.keys(featureAttributes).length;

  return (
    <ContentPanel
      title="Model configuration"
      titleSize="s"
      actions={[
        <EuiSmallButton onClick={props.onEditModelConfiguration}>Edit</EuiSmallButton>,
      ]}
    >
      {getValidationCallout()}

      <div>
        <AdditionalSettings
          shingleSize={shingleSize}
          categoryField={get(props, 'detector.categoryField', [])}
          imputationMethod={imputationMethodStr}
          customValues={getCustomValueStrArray(imputationMethodStr, props.detector)}
        />
        <EuiSpacer />
        <ContentPanel
          title={getTitleWithCount('Features', featureNum)}
          titleSize="s"
        >
          <EuiBasicTable
            data-test-subj="featureTable"
            items={sortedItems}
            columns={columns}
            cellProps={getCellProps}
            sorting={sorting}
            onChange={handleTableChange}
          />
        </ContentPanel>
      </div>
    </ContentPanel>
  );
};
