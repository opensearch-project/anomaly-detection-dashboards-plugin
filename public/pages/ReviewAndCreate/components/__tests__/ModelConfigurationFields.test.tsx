/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import chance from 'chance';
import userEvent from '@testing-library/user-event';
import { render, waitFor, fireEvent, screen, within } from '@testing-library/react';
import { ModelConfigurationFields } from '../ModelConfigurationFields/ModelConfigurationFields';
import {
  Detector,
  UNITS,
  FILTER_TYPES,
  OPERATORS_MAP,
  ValidationModelResponse,
} from '../../../../models/interfaces';
import { DATA_TYPES } from '../../../../utils/constants';
import { getRandomFeature } from '../../../../redux/reducers/__tests__/utils';
import { CoreServicesContext } from '../../../../components/CoreServices/CoreServices';
import { coreServicesMock } from '../../../../../test/mocks';
import {
  ImputationMethod,
  ThresholdType,
  Action,
  Operator,
  Rule
} from '../../../../models/types';

const detectorFaker = new chance('seed');
const features = new Array(detectorFaker.natural({ min: 1, max: 5 }))
  .fill(null)
  .map(() => getRandomFeature(false));

// Generate rules based on the existing features
const rules = features.map((feature, index) => ({
  action: Action.IGNORE_ANOMALY,
  conditions: [
    {
      featureName: feature.featureName,
      thresholdType: index % 2 === 0 ? ThresholdType.ACTUAL_OVER_EXPECTED_MARGIN : ThresholdType.EXPECTED_OVER_ACTUAL_RATIO, // Alternate threshold types for variety
      operator: Operator.LTE,
      value: index % 2 === 0 ? 5 : 0.1, // Use different values for variety
    },
  ],
})) as Rule[];

const testDetector = {
  id: 'test-id',
  name: 'test-detector',
  indices: ['test-index'],
  detectionInterval: {
    period: {
      interval: 10,
      unit: UNITS.MINUTES,
    },
  },
  description: 'test-description',
  timeField: 'test-timefield',
  windowDelay: {
    period: {
      interval: 1,
      unit: UNITS.MINUTES,
    },
  },
  uiMetadata: {
    filters: [
      {
        filterType: FILTER_TYPES.SIMPLE,
        fieldInfo: [
          {
            label: 'test-filter-field',
            type: DATA_TYPES.TEXT,
          },
        ],
        operator: OPERATORS_MAP.IS,
        fieldValue: 'null',
      },
    ],
  },
  featureAttributes: features,
  imputationOption: { method: ImputationMethod.ZERO},
  rules: rules
} as Detector;

describe('ModelConfigurationFields', () => {
  test('renders the component in create mode (no ID)', async () => {
    const onEditModelConfiguration = jest.fn();
    const { container, getByText, getByTestId, queryByText, getByRole, queryByRole } = render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        <ModelConfigurationFields
          detector={testDetector}
          onEditModelConfiguration={onEditModelConfiguration}
          validationFeatureResponse={{} as ValidationModelResponse}
          validModel={true}
          validationError={false}
          isLoading={false}
          isCreatingDetector={true}
        />
      </CoreServicesContext.Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
    getByText('set_to_zero');

    // Check for the suppression rules button link
    const button = getByRole('button', { name: '2 rules' });
    expect(button).toBeInTheDocument();

    userEvent.click(getByTestId('viewFeature-0'));
    await waitFor(() => {
      queryByText('max');
    });
  });
});
