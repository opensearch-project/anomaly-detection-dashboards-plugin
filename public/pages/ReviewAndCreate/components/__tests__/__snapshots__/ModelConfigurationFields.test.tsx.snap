// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelConfigurationFields renders the component in create mode (no ID) 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
  style="padding: 20px;"
>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
    style="padding: 0px;"
  >
    <div
      class="euiFlexItem"
    >
      <h3
        class="euiTitle euiTitle--small"
        data-test-subj="contentPanelTitle"
      >
        Model configuration
      </h3>
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem content-panel-subTitle"
          style="line-height: normal; max-width: 75%;"
        />
      </div>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem"
        >
          <button
            class="euiButton euiButton--primary euiButton--small"
            type="button"
          >
            <span
              class="euiButtonContent euiButton__content"
            >
              <span
                class="euiButton__text"
              >
                Edit
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--s"
  />
  <div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
    />
    <div
      style="padding: 10px 0px;"
    >
      <div
        class="euiCallOut euiCallOut--success euiCallOut--small"
        style="margin-bottom: 10px;"
      >
        <div
          class="euiCallOutHeader"
        >
          <svg
            aria-hidden="true"
            class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiCallOutHeader__icon"
            focusable="false"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
            />
          </svg>
          <span
            class="euiCallOutHeader__title"
          >
            Model configurations are validated
          </span>
        </div>
      </div>
      <div>
        <div
          class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
          style="padding: 20px;"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
            style="padding: 0px;"
          >
            <div
              class="euiFlexItem"
            >
              <h3
                class="euiTitle euiTitle--small"
                data-test-subj="contentPanelTitle"
              >
                Additional settings
              </h3>
              <div
                class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
              >
                <div
                  class="euiFlexItem content-panel-subTitle"
                  style="line-height: normal; max-width: 75%;"
                />
              </div>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <div
                class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
              >
                <div
                  class="euiFlexItem"
                />
              </div>
            </div>
          </div>
          <div>
            <hr
              class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
            />
            <div
              style="padding: 10px 0px;"
            >
              <div
                class="euiBasicTable header-single-value-euiBasicTable"
              >
                <div>
                  <div
                    class="euiTableHeaderMobile"
                  >
                    <div
                      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
                    >
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                      />
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                      />
                    </div>
                  </div>
                  <table
                    class="euiTable euiTable--responsive"
                    id="random_html_id"
                    tabindex="-1"
                  >
                    <caption
                      class="euiScreenReaderOnly euiTableCaption"
                    />
                    <thead>
                      <tr>
                        <th
                          class="euiTableHeaderCell"
                          data-test-subj="tableHeaderCell_categoryField_0"
                          role="columnheader"
                          scope="col"
                        >
                          <span
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                              title="Category field"
                            >
                              Category field
                            </span>
                          </span>
                        </th>
                        <th
                          class="euiTableHeaderCell"
                          data-test-subj="tableHeaderCell_shingleSize_1"
                          role="columnheader"
                          scope="col"
                        >
                          <span
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                              title="Shingle size"
                            >
                              Shingle size
                            </span>
                          </span>
                        </th>
                        <th
                          class="euiTableHeaderCell"
                          data-test-subj="tableHeaderCell_imputationMethod_2"
                          role="columnheader"
                          scope="col"
                        >
                          <span
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                              title="Imputation method"
                            >
                              Imputation method
                            </span>
                          </span>
                        </th>
                        <th
                          class="euiTableHeaderCell"
                          data-test-subj="tableHeaderCell_customValues_3"
                          role="columnheader"
                          scope="col"
                        >
                          <span
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                              title="Custom values"
                            >
                              Custom values
                            </span>
                          </span>
                        </th>
                        <th
                          class="euiTableHeaderCell"
                          data-test-subj="tableHeaderCell_suppresionRules_4"
                          role="columnheader"
                          scope="col"
                        >
                          <span
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                              title="Suppression rules"
                            >
                              Suppression rules
                            </span>
                          </span>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr
                        class="euiTableRow"
                      >
                        <td
                          class="euiTableRowCell"
                        >
                          <div
                            class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                          >
                            Category field
                          </div>
                          <div
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                            >
                              -
                            </span>
                          </div>
                        </td>
                        <td
                          class="euiTableRowCell"
                        >
                          <div
                            class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                          >
                            Shingle size
                          </div>
                          <div
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                            >
                              8
                            </span>
                          </div>
                        </td>
                        <td
                          class="euiTableRowCell"
                        >
                          <div
                            class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                          >
                            Imputation method
                          </div>
                          <div
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                            >
                              set_to_zero
                            </span>
                          </div>
                        </td>
                        <td
                          class="euiTableRowCell"
                        >
                          <div
                            class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                          >
                            Custom values
                          </div>
                          <div
                            class="euiTableCellContent euiTableCellContent--overflowingContent"
                          >
                            <div
                              class=""
                            >
                              <p>
                                -
                              </p>
                            </div>
                          </div>
                        </td>
                        <td
                          class="euiTableRowCell"
                        >
                          <div
                            class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                          >
                            Suppression rules
                          </div>
                          <div
                            class="euiTableCellContent euiTableCellContent--overflowingContent"
                          >
                            <div
                              class=""
                            >
                              <button
                                class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--small"
                                type="button"
                              >
                                <span
                                  class="euiButtonContent euiButtonEmpty__content"
                                >
                                  <span
                                    class="euiButtonEmpty__text"
                                  >
                                    2
                                     rules
                                  </span>
                                </span>
                              </button>
                            </div>
                          </div>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiSpacer euiSpacer--l"
        />
        <div
          class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
          style="padding: 20px;"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
            style="padding: 0px;"
          >
            <div
              class="euiFlexItem"
            >
              <div
                class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
                data-test-subj="contentPanelTitle"
              >
                <div
                  class="euiFlexItem"
                >
                  <h3
                    class="euiTitle euiTitle--small"
                    style="display: flex; flex-direction: row;"
                  >
                    <p>
                      Features
                       
                    </p>
                    <p
                      style="color: rgb(83, 89, 102);"
                    >
                      (2)
                    </p>
                  </h3>
                </div>
              </div>
              <div
                class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
              >
                <div
                  class="euiFlexItem content-panel-subTitle"
                  style="line-height: normal; max-width: 75%;"
                />
              </div>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <div
                class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
              >
                <div
                  class="euiFlexItem"
                />
              </div>
            </div>
          </div>
          <div>
            <hr
              class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
            />
            <div
              style="padding: 10px 0px;"
            >
              <div
                class="euiBasicTable"
                data-test-subj="featureTable"
              >
                <div>
                  <div
                    class="euiTableHeaderMobile"
                  >
                    <div
                      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
                    >
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                      />
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                      >
                        <div
                          class="euiTableSortMobile"
                        >
                          <div
                            class="euiPopover euiPopover--anchorDownRight"
                          >
                            <div
                              class="euiPopover__anchor"
                            >
                              <button
                                class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiButtonEmpty--flushRight"
                                type="button"
                              >
                                <span
                                  class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                                >
                                  <svg
                                    aria-hidden="true"
                                    class="euiIcon euiIcon--small euiIcon--inherit euiIcon-isLoading euiButtonContent__icon"
                                    focusable="false"
                                    height="16"
                                    role="img"
                                    viewBox="0 0 16 16"
                                    width="16"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <path
                                      d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                                    />
                                  </svg>
                                  <span
                                    class="euiButtonEmpty__text"
                                  >
                                    Sorting
                                  </span>
                                </span>
                              </button>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <table
                    class="euiTable euiTable--responsive"
                    id="random_html_id"
                    tabindex="-1"
                  >
                    <caption
                      class="euiScreenReaderOnly euiTableCaption"
                    />
                    <thead>
                      <tr>
                        <th
                          aria-live="polite"
                          aria-sort="ascending"
                          class="euiTableHeaderCell"
                          data-test-subj="tableHeaderCell_name_0"
                          role="columnheader"
                          scope="col"
                        >
                          <button
                            class="euiTableHeaderButton euiTableHeaderButton-isSorted"
                            data-test-subj="tableHeaderSortButton"
                            type="button"
                          >
                            <span
                              class="euiTableCellContent"
                            >
                              <span
                                class="euiTableCellContent__text"
                                title="Feature name"
                              >
                                Feature name
                              </span>
                              <svg
                                aria-hidden="true"
                                class="euiIcon euiIcon--medium euiIcon-isLoading euiTableSortIcon"
                                focusable="false"
                                height="16"
                                role="img"
                                viewBox="0 0 16 16"
                                width="16"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                                />
                              </svg>
                            </span>
                          </button>
                        </th>
                        <th
                          class="euiTableHeaderCell"
                          data-test-subj="tableHeaderCell_definition_1"
                          role="columnheader"
                          scope="col"
                        >
                          <span
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                              title="Feature definition"
                            >
                              Feature definition
                            </span>
                          </span>
                        </th>
                        <th
                          class="euiTableHeaderCell"
                          data-test-subj="tableHeaderCell_state_2"
                          role="columnheader"
                          scope="col"
                        >
                          <span
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                              title="Feature state"
                            >
                              Feature state
                            </span>
                          </span>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr
                        class="euiTableRow"
                      >
                        <td
                          class="euiTableRowCell"
                        >
                          <div
                            class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                          >
                            Feature name
                          </div>
                          <div
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                            >
                              TF
                            </span>
                          </div>
                        </td>
                        <td
                          class="euiTableRowCell"
                        >
                          <div
                            class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                          >
                            Feature definition
                          </div>
                          <div
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                            >
                              <div>
                                <p>
                                  Custom expression:
                                   
                                  <button
                                    class="euiLink euiLink--primary"
                                    data-test-subj="viewFeature-0"
                                    type="button"
                                  >
                                    View code
                                  </button>
                                </p>
                              </div>
                            </span>
                          </div>
                        </td>
                        <td
                          class="euiTableRowCell"
                        >
                          <div
                            class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                          >
                            Feature state
                          </div>
                          <div
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                            >
                              Disabled
                            </span>
                          </div>
                        </td>
                      </tr>
                      <tr
                        class="euiTableRow"
                      >
                        <td
                          class="euiTableRowCell"
                        >
                          <div
                            class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                          >
                            Feature name
                          </div>
                          <div
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                            >
                              TR
                            </span>
                          </div>
                        </td>
                        <td
                          class="euiTableRowCell"
                        >
                          <div
                            class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                          >
                            Feature definition
                          </div>
                          <div
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                            >
                              <div>
                                <p>
                                  Custom expression:
                                   
                                  <button
                                    class="euiLink euiLink--primary"
                                    data-test-subj="viewFeature-1"
                                    type="button"
                                  >
                                    View code
                                  </button>
                                </p>
                              </div>
                            </span>
                          </div>
                        </td>
                        <td
                          class="euiTableRowCell"
                        >
                          <div
                            class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                          >
                            Feature state
                          </div>
                          <div
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                            >
                              Disabled
                            </span>
                          </div>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
